darkcluster/src/main/java/com/linkedin/darkcluster/util/ExpiringCircularBuffer.java

```diff
@@ -0,0 +1,143 @@
+/*
+   Copyright (c) 2021 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.darkcluster.util;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Queue;
+import org.apache.commons.lang3.NotImplementedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * TODO: javadocs
+ * @param <E>
+ */
+@SuppressWarnings("serial")
+public class ExpiringCircularBuffer<E> extends ArrayList<E> implements Queue<E> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ExpiringCircularBuffer.class);
+  private Duration _ttl;
+  private ArrayList<Instant> _ttlBuffer;
+  private volatile int _readerPosition;
+  private volatile int _writerPosition;
+
+  public ExpiringCircularBuffer(int capacity, int ttl, ChronoUnit ttlUnit) {
+    super();
+    setCapacity(capacity);
+    setTtl(ttl, ttlUnit);
+  }
+
+  public synchronized boolean add(E toAdd) {
+    this.set(_writerPosition, toAdd);
+    _ttlBuffer.set(_writerPosition, Instant.now());
+    bumpWriterPosition();
+    return true;
+  }
+
+  /**
+   * TODO: javadocs
+   * @return
+   */
+  public synchronized E element() {
```

Comment:
Can we have a more granularized synchronization control using a private lock object, instead of method-level synchronization? You can take a look at: https://www.baeldung.com/java-synchronization-bad-practices. I would imagine for most parts of the read operation, we don't need to obtain the lock.

Reply:
Yeah, I can do better ;)